File: packages/xpute/core/src/collection/deque.ts
// @xpute/core/collection/deque.ts
// deque primitives over an external ring buffer (0-based) — DOD view.
//
// Storage contract (caller-owned):
//   - buf.length is the fixed capacity (cap)
//   - head: pop-front index
//   - len : live element count
//
// No defensive checks. Caller upholds preconditions.

import type { u32 } from "@xpute/core/abi/word.ts";

export class Deque<T> {
  // external storage (DOD-friendly): caller owns the ring buffer.
  constructor(
    public readonly buf: (T | undefined)[],
    public head: u32 = 0,
    public len: u32 = 0,
  ) {}

  cap(): u32 {
    return this.buf.length as u32;
  }

  size(): u32 {
    return this.len;
  }

  empty(): boolean {
    return this.len === 0;
  }

  full(): boolean {
    return this.len === (this.buf.length as u32);
  }

  front(): T | undefined {
    return this.len ? (this.buf[this.head] as T) : undefined;
  }

  back(): T | undefined {
    if (!this.len) return undefined;
    const cap = this.cap();
    const i = (this.head + this.len - 1) % cap;
    return this.buf[i] as T;
  }

  push_back(v: T): void {
    const cap = this.cap();
    const i = (this.head + this.len) % cap;
    this.buf[i] = v;
    this.len++;
  }

  push_front(v: T): void {
    const cap = this.cap();
    this.head = (this.head + cap - 1) % cap;
    this.buf[this.head] = v;
    this.len++;
  }

  pop_front(): T {
    const v = this.buf[this.head] as T;
    this.buf[this.head] = undefined;
    this.head = (this.head + 1) % this.cap();
    this.len--;
    return v;
  }

  pop_back(): T {
    const cap = this.cap();
    const i = (this.head + this.len - 1) % cap;
    const v = this.buf[i] as T;
    this.buf[i] = undefined;
    this.len--;
    return v;
  }

  clear(): void {
    const cap = this.cap();
    for (let i = 0; i < this.len; i++) this.buf[(this.head + i) % cap] = undefined;
    this.head = 0;
    this.len = 0;
  }
}

File: crates/xpute/core/src/collection/deque.rs
// xpute-core/collection/deque.rs

//! A ring over a buffer the caller owns: `buf` is the capacity, `head` is
//! where the front sits, and `len` counts live elements forward from there.
//! The three are public because they are the caller's — this allocates
//! nothing, grows never, and holds no state of its own.
//!
//! A slot is `Option<T>` because that is how a value leaves a borrowed slice
//! without unsafe. It carries no meaning: `head` and `len` alone say which
//! slots are live.

pub struct Deque<'a, T> {
    pub buf: &'a mut [Option<T>],
    pub head: u32,
    pub len: u32,
}

impl<'a, T> Deque<'a, T> {
    pub fn new(buf: &'a mut [Option<T>], head: u32, len: u32) -> Deque<'a, T> {
        Deque { buf, head, len }
    }
}

impl<T> Deque<'_, T> {
    pub fn cap(&self) -> u32 {
        self.buf.len() as u32
    }

    pub fn is_empty(&self) -> bool {
        self.len == 0
    }

    pub fn is_full(&self) -> bool {
        self.len == self.cap()
    }

    pub fn front(&self) -> Option<&T> {
        if self.len == 0 {
            return None;
        }
        self.buf[self.head as usize].as_ref()
    }

    pub fn back(&self) -> Option<&T> {
        if self.len == 0 {
            return None;
        }
        let i = (self.head + self.len - 1) % self.cap();
        self.buf[i as usize].as_ref()
    }

    pub fn push_back(&mut self, v: T) {
        let i = (self.head + self.len) % self.cap();
        self.buf[i as usize] = Some(v);
        self.len += 1;
    }

    pub fn push_front(&mut self, v: T) {
        let cap = self.cap();
        self.head = (self.head + cap - 1) % cap;
        self.buf[self.head as usize] = Some(v);
        self.len += 1;
    }

    pub fn pop_front(&mut self) -> Option<T> {
        if self.len == 0 {
            return None;
        }
        let v = self.buf[self.head as usize].take();
        self.head = (self.head + 1) % self.cap();
        self.len -= 1;
        v
    }

    pub fn pop_back(&mut self) -> Option<T> {
        if self.len == 0 {
            return None;
        }
        let i = (self.head + self.len - 1) % self.cap();
        self.len -= 1;
        self.buf[i as usize].take()
    }

    pub fn clear(&mut self) {
        let cap = self.cap();
        for i in 0..self.len {
            self.buf[((self.head + i) % cap) as usize] = None;
        }
        self.head = 0;
        self.len = 0;
    }
}

#[cfg(test)]
#[path = "deque.test.rs"]
mod test;
